src/fdcache_entity.cpp

```diff
@@ -1489,6 +1489,15 @@ ssize_t FdEntity::Write(const char* bytes, off_t start, size_t size)
                 S3FS_PRN_WARN("Not enough local storage to cache write request: [path=%s][fd=%d][offset=%lld][size=%zu]", path.c_str(), fd, static_cast<long long int>(start), size);
                 return -ENOSPC;   // No space left on device
             }
+            if (start < MIN_MULTIPART_SIZE) {
+                S3FS_PRN_WARN("Not enough local storage to cache write request till multipart upload can start: [path=%s][fd=%d][offset=%lld][size=%zu]", path.c_str(), fd, static_cast<long long int>(start), size);
+                return -ENOSPC;   // No space left on device
+            }
+            if (start < S3fsCurl::GetMultipartSize()) {
+                // ensure that all parts have the same size
+                S3FS_PRN_WARN("Not enough local storage to fully cache write request, reduce multipart size permanently from %lld to %lld to start upload: [path=%s][fd=%d][offset=%lld][size=%zu]", static_cast<long long int>(S3fsCurl::GetMultipartSize()), static_cast<long long int>(start), path.c_str(), fd, static_cast<long long int>(start), size);
+                S3fsCurl::SetMultipartSize(start);
+            }
```

Comment:
I am concerned that the existing logic is untested and complicated enough that I don't have confidence in it.  While the suggested fix seems plausible, can we step back and evaluate what the ideal logic would look like?  I think if the user lacks sufficient space other code path are likely to fail.

@ggtakec Can we just return `ENOSPC` and simplify this?

Reply:
@CarstenGrohmann Thanks for your help.

I think the process of (1) is appropriate.
Regarding part (2), I feel the danger of dynamically resizing.
One of the reasons is that there is no chance to restore the size once changed.
Instead of resizing(case 2), if `start + size` is less than or equal to the multipart size, I recommend that it returns an ENOSPC or does not start the multipart upload.
(Even if it don't return ENOSPC at here, this Write method will be called repeatedly and it will eventually get an error.)

I think that in both cases, user should essentially reserve the free space indicated by ensure_diskfree and specify the size of the multipart upload commensurate with that size.

@gaul I think that preparing a test case for this fix can be difficult.
If we do not set(know) the free space accurately, we will not be able to create test cases well.

Reply:
@ggtakec I've incorperated the suggested changes.

Regarding a test case, you mean something like this (untested):
```
dd if=/dev/urandom of="${TEMP_DIR}/${BIG_FILE}" bs=$BIG_FILE_BLOCK_SIZE count=$BIG_FILE_COUNT

remaining_kb=1024
free_kb=$(df -k /tmp/ | awk 'NR == 2 {print $4}')
to_fill=$(( $free_kb - $remaining_kb ))
tmp_zero="/tmp/zero.$$"
dd if=/dev/zero of=$tmp_zero bs=1k count=$to_fill

(
    set +o pipefail
    cp ${TEMP_DIR}/${BIG_FILE} ${BIG_FILE} 2>&1 | grep -q "No space left on device"
)
rm $tmp_zero
m -f ${TEMP_DIR}/${BIG_FILE}
rm_test_file ${BIG_FILE}
```

Reply:
@CarstenGrohmann Thanks, it looks good to me.
@gaul Please give us your opinion.